nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/cosmos/document/AzureCosmosDocumentConnectionControllerService.java

```diff
@@ -0,0 +1,192 @@
+import java.util.List;
+
+import com.azure.cosmos.ConnectionPolicy;
+import com.azure.cosmos.ConsistencyLevel;
+import com.azure.cosmos.CosmosClient;
+import com.azure.cosmos.CosmosClientBuilder;
+
+@Tags({"azure", "cosmos", "document", "service"})
+@CapabilityDescription(
+        "Provides a controller service that configures a connection to Azure Cosmos Document DB (Document API or Core SQL API recently renamed) " +
+                " and provides access to that connection to other CosmosDB-related components."
+)
+public class AzureCosmosDocumentConnectionControllerService extends AbstractControllerService implements AzureCosmosDocumentConnectionService {
+    static final String CONSISTENCY_STRONG = "STRONG";
+    static final String CONSISTENCY_BOUNDED_STALENESS= "BOUNDED_STALENESS";
+    static final String CONSISTENCY_SESSION = "SESSION";
+    static final String CONSISTENCY_CONSISTENT_PREFIX = "CONSISTENT_PREFIX";
+    static final String CONSISTENCY_EVENTUAL = "EVENTUAL";
+    private String uri;
+    private String accessKey;
+    protected CosmosClient cosmosClient;
+
+    public static final PropertyDescriptor URI = new PropertyDescriptor.Builder()
+    .name("Cosmos Document DB URI")
+    .displayName("Cosmos Document DB URI")
+    .description("CosmosURI, typically of the form: https://{databaseaccount}.documents.azure.com:443/. Note this URI is for Cosmos DB with SQL API")
+    .required(true)
+    .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+    .build();
+
+    public static final PropertyDescriptor DB_ACCESS_KEY = new PropertyDescriptor.Builder()
+        .name("Cosmos Document DB Access Key")
+        .displayName("Cosmos Document DB Access Key")
+        .description("Cosmos DB Access Key from Azure Portal (Settings->Keys)")
+        .required(true)
+        .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+        .sensitive(true)
+        .build();
+
+    public static final PropertyDescriptor CONSISTENCY = new PropertyDescriptor.Builder()
+        .name("Cosmos Consistency Level")
+        .displayName("Cosmos Consistency Level")
+        .description("Cosmos Consistency Level to use")
+        .required(false)
+        .allowableValues(CONSISTENCY_STRONG, CONSISTENCY_BOUNDED_STALENESS, CONSISTENCY_SESSION,
+                CONSISTENCY_CONSISTENT_PREFIX, CONSISTENCY_EVENTUAL)
+        .defaultValue(CONSISTENCY_SESSION)
+        .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+        .build();
+
+    @OnEnabled
+    public void onEnabled(final ConfigurationContext context) {
+        this.uri = context.getProperty(URI).getValue();
+        this.accessKey = context.getProperty(DB_ACCESS_KEY).getValue();
+        final ConsistencyLevel clevel;
+        final String selectedConsistency;
+        if(context.getProperty(CONSISTENCY).isSet()){
+            selectedConsistency = context.getProperty(CONSISTENCY).getValue();
+        }else {
+            selectedConsistency =  CONSISTENCY_SESSION;
```

Comment:
I'm not sure I understand this. Why is session used consistency is not set?

Reply:
Since I set the default value for CONSISTENCY as CONSISTENCY_SESSION, this is redundant. I will take it out.

Reply:
It looks like its' being set in the CASE statement as well below. My question is a bit different - you can define a default "consistency level" for a CosmosDB on the service itself. So if select "No Value" for the consistency level in the processor should I expect the client to use the default set on the service instance?

Reply:
And, setting the DB connection configuration and consistency has to be done either thru a controller or processor. It is mutually exclusive. My logic is in customValidate. 

And I was mentioning default value setting with PropertyDescriptor.
 
    static final PropertyDescriptor CONSISTENCY = new PropertyDescriptor.Builder()
        .name("Azure Cosmos Consistency Level")
        .displayName("Azure Cosmos Consistency Level")
        .description("Azure Cosmos Consistency Level to use")
        .required(false)
        .allowableValues(CONSISTENCY_STRONG, CONSISTENCY_BOUNDED_STALENESS, CONSISTENCY_SESSION,
                CONSISTENCY_CONSISTENT_PREFIX, CONSISTENCY_EVENTUAL)
        .defaultValue(CONSISTENCY_SESSION)
        .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
        .build();

Reply:
It turns out I have to check if context.getProperty(CONSISTENCY).isSet() for unit tests. 

CONSISTENCY is not a required field, with the default value in the PropertyDescriptor. In the flow design, the default value is picked up and configured. However, in the NIFI unit tests, this default value is not set and thus it returns null value, because it is not required field.. So, I need to keep the logic.

Reply:
And, DB connection configuration method is exclusive: either using controller service or processors properties, so consistency level can not be picked up from controller service by processor instance.